Gordon360/Services/EventService.cs

```diff
@@ -63,6 +63,39 @@ public IEnumerable<EventViewModel> GetCLAWEvents()
         return Events.Where(e => e.HasCLAWCredit).OrderBy(e => e.StartDate);
     }
 
+    /// <summary>
+    /// Select only events that are Final Exams for users' all courses
+    /// </summary>
+    /// <param name="username"> The student's AD Username</param>
+    /// <returns>All Final Exam Events for the user</returns>
+    public IEnumerable<EventViewModel> GetFinalExamEventsForUser(string username)
+    {
+        // Get all courses for the user
+        var coursesBySession = scheduleService.GetAllCoursesAsync(username).GetAwaiter().GetResult();
+        var userCourseCodes = coursesBySession
+            .SelectMany(session => session.AllCourses)
+            .Select(course => NormalizeSpaces(course.CRS_CDE))
+            .ToHashSet(StringComparer.OrdinalIgnoreCase);
```

Comment:
This is a really wasteful way to get the course code for all the courses the user is in. Especially since we really only care about the courses for one term (the term that we're getting final exams for).

The better solution would be to query the `ScheduleCourse` view and select distinct course codes, ideally with a where filter for only one term's courses.